sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2178,26 +2178,54 @@ std::pair<RT::PiKernel, std::mutex *> ProgramManager::getOrCreateKernel(
                         &(BuildResult->MBuildResultMutex));
 }
 
-bool doesDevSupportImgAspects(const device &Dev,
-                              const RTDeviceBinaryImage &Img) {
-  const RTDeviceBinaryImage::PropertyRange &PropRange =
-      Img.getDeviceRequirements();
-  RTDeviceBinaryImage::PropertyRange::ConstIterator PropIt = std::find_if(
-      PropRange.begin(), PropRange.end(),
-      [](RTDeviceBinaryImage::PropertyRange::ConstIterator &&Prop) {
-        using namespace std::literals;
-        return (*Prop)->Name == "aspects"sv;
-      });
-  if (PropIt == PropRange.end())
+bool doesDevSupportDeviceRequirements(const device &Dev,
+                                      const RTDeviceBinaryImage &Img) {
+  auto getPropValue = [&Img](std::string PropName) {
+    const RTDeviceBinaryImage::PropertyRange &PropRange =
+        Img.getDeviceRequirements();
+    RTDeviceBinaryImage::PropertyRange::ConstIterator PropIt = std::find_if(
+        PropRange.begin(), PropRange.end(),
+        [&PropName](RTDeviceBinaryImage::PropertyRange::ConstIterator &&Prop) {
+          return (*Prop)->Name == PropName;
+        });
+    bool IsRangeEnd = false;
+    if (PropIt == PropRange.end())
+      IsRangeEnd = true;
+    return std::make_pair(IsRangeEnd, *PropIt);
+  };
+
+  enum { IS_RANGE_END, PROP_VAL };
+
+  // Checking if device supports defined aspects
+  auto AspectsPropVal = getPropValue("aspects");
+  if (std::get<IS_RANGE_END>(AspectsPropVal))
     return true;
-  ByteArray Aspects = DeviceBinaryProperty(*PropIt).asByteArray();
+  ByteArray Aspects =
+      DeviceBinaryProperty(std::get<PROP_VAL>(AspectsPropVal)).asByteArray();
   // Drop 8 bytes describing the size of the byte array.
   Aspects.dropBytes(8);
   while (!Aspects.empty()) {
     aspect Aspect = Aspects.consume<aspect>();
     if (!Dev.has(Aspect))
       return false;
   }
+
+  // Checking if device supports defined required work group size
+  auto ReqdWGSizePropVal = getPropValue("reqd_work_group_size");
+  if (std::get<IS_RANGE_END>(ReqdWGSizePropVal))
+    return true;
+  ByteArray ReqdWGSize =
+      DeviceBinaryProperty(std::get<PROP_VAL>(ReqdWGSizePropVal)).asByteArray();
+  // Drop 8 bytes describing the size of the byte array.
+  ReqdWGSize.dropBytes(8);
+  int ReqdWGSizeAllDimsSum = 0;
+  while (!ReqdWGSize.empty()) {
+    ReqdWGSizeAllDimsSum += ReqdWGSize.consume<int>();
```

Comment:
It should be a multiplication, not a sum. `reqd_work_group_size(4, 4, 4)` describes a 3-dimensional work-group with size of each side equal to 4, i.e. it contains 64 work-items, not 12.